src/Maestro/FeedCleaner/FeedCleaner.cs

```diff
@@ -0,0 +1,369 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Maestro.AzureDevOps;
+using Maestro.Contracts;
+using Maestro.Data;
+using Maestro.Data.Models;
+using Microsoft.DotNet.DarcLib;
+using Microsoft.DotNet.ServiceFabric.ServiceHost;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace FeedCleaner
+{
+
+    /// <summary>
+    ///     An instance of this class is created for each service replica by the Service Fabric runtime.
+    /// </summary>
+    public sealed class FeedCleaner : IFeedCleaner, IServiceImplementation
+    {
+
+        private readonly string MaestroManagedFeedPattern = @"darc-(?<type>(int|pub))-(?<repository>.+?)-(?<sha>[A-Fa-f0-9]{7,40})-?(?<subversion>\d*)";
+        private static readonly string NuGetOrgRegistrationBaseUrl = "https://api.nuget.org/v3/registration3-gz-semver2";
+        private static readonly string NuGetOrgLocation = "https://api.nuget.org/v3/index.json";
+
+        public FeedCleaner(
+            ILogger<FeedCleaner> logger,
+            BuildAssetRegistryContext context,
+            IOptions<FeedCleanerOptions> options)
+        {
+            Logger = logger;
+            Context = context;
+            _httpClient = new HttpClient();
+            _options = options;
+            AzureDevOpsClients = new Dictionary<string, IAzureDevOpsClient>();
+            foreach (string account in _options.Value.AzdoAccounts)
+            {
+                AzureDevOpsClients.Add(account, GetAzureDevOpsClientForAccountAsync(account));
+            }
+        }
+
+        public ILogger<FeedCleaner> Logger { get; }
+        public BuildAssetRegistryContext Context { get; }
+
+        public FeedCleanerOptions Options => _options.Value;
+
+        public Dictionary<string, IAzureDevOpsClient> AzureDevOpsClients { get; set; }
+
+        private readonly HttpClient _httpClient;
+        private readonly IOptions<FeedCleanerOptions> _options;
+
+        public Task<TimeSpan> RunAsync(CancellationToken cancellationToken)
+        {
+            return Task.FromResult(TimeSpan.FromMinutes(5));
+        }
+
+        /// <summary>
+        /// Updates assets that are now available from one of the non-stable feeds,
+        /// delete those package versions from the stable feeds and delete any feeds 
+        /// where all packages versions have been deleted every day at 2 AM.
+        /// </summary>
+        [CronSchedule("0 0 2 1/1 * ? *", TimeZones.PST)]
+        public async Task CleanManagedFeedsAsync()
+        {
+            if (Options.Enabled)
+            {
+                Dictionary<string, Dictionary<string, HashSet<string>>> packagesInReleaseFeeds =
+                    await GetPackagesForReleaseFeedsAsync();
```

Comment:
Is this a particularly expensive operation? It seems like the number of packages on the release feeds will steadily increase over time, while the number of packages in the managed feeds will increase in short bursts, then decrease after release day.

Reply:
This is a computationally expensive operation to generate the mappings for lookups, but this is just two API calls per release feed, one to get the feed information, and another to get all packages and their versions from said feed.